modules/control/submodules/preprocessor_submodule.cc

```diff
@@ -0,0 +1,333 @@
+        first_trajectory_point.v() < -1 * kEpsilon) {
+      estop_ = true;
+      estop_reason_ = "estop for negative speed when gear_drive";
+    }
+  }
+
+  if (!estop_) {
+    auto debug = preprocessor_status->mutable_input_debug();
+    debug->mutable_localization_header()->CopyFrom(
+        local_view_.localization.header());
+    debug->mutable_canbus_header()->CopyFrom(local_view_.chassis.header());
+    debug->mutable_trajectory_header()->CopyFrom(
+        local_view_.trajectory.header());
+
+    if (local_view_.trajectory.is_replan()) {
+      latest_replan_trajectory_header_.CopyFrom(
+          local_view_.trajectory.header());
+    }
+
+    if (latest_replan_trajectory_header_.has_sequence_num()) {
+      debug->mutable_latest_replan_trajectory_header()->CopyFrom(
+          latest_replan_trajectory_header_);
+    }
+  }
+
+  return status;
+}
+
+void PreprocessorSubmodule::OnChassis(const std::shared_ptr<Chassis> &chassis) {
+  ADEBUG << "Received chassis data: run chassis callback.";
+  std::lock_guard<std::mutex> lock(mutex_);
+  latest_chassis_.CopyFrom(*chassis);
+}
+
+void PreprocessorSubmodule::OnPad(const std::shared_ptr<PadMessage> &pad) {
+  pad_msg_.CopyFrom(*pad);
+  ADEBUG << "Received Pad Msg:" << pad_msg_.DebugString();
+  AERROR_IF(!pad_msg_.has_action()) << "pad message check failed!";
+
+  if (pad_msg_.action() == DrivingAction::RESET) {
+    AINFO << "Control received RESET action!";
+    estop_ = false;
+    estop_reason_.clear();
+  }
+  pad_received_ = true;
+}
+
+void PreprocessorSubmodule::OnPlanning(
+    const std::shared_ptr<ADCTrajectory> &trajectory) {
+  ADEBUG << "Received chassis data: run trajectory callback.";
+  std::lock_guard<std::mutex> lock(mutex_);
+  latest_trajectory_.CopyFrom(*trajectory);
+}
+
+void PreprocessorSubmodule::OnLocalization(
+    const std::shared_ptr<LocalizationEstimate> &localization) {
+  ADEBUG << "Received control data: run localization message callback.";
+  std::lock_guard<std::mutex> lock(mutex_);
+  latest_localization_.CopyFrom(*localization);
+}
+
+void PreprocessorSubmodule::OnMonitor(
+    const common::monitor::MonitorMessage &monitor_message) {
+  for (const auto &item : monitor_message.item()) {
+    if (item.log_level() == common::monitor::MonitorMessageItem::FATAL) {
+      estop_ = true;
+      return;
+    }
+  }
+}
+
+Status PreprocessorSubmodule::CheckInput(LocalView *local_view) {
+  ADEBUG << "Received localization:"
+         << local_view->localization.ShortDebugString();
+  ADEBUG << "Received chassis:" << local_view->chassis.ShortDebugString();
+
+  if (!local_view->trajectory.estop().is_estop() &&
+      local_view->trajectory.trajectory_point().empty()) {
+    AWARN_EVERY(100) << "planning has no trajectory point. ";
+    std::string msg("planning has no trajectory point. planning_seq_num:");
+    msg += std::to_string(local_view->trajectory.header().sequence_num());
+    return Status(ErrorCode::CONTROL_COMPUTE_ERROR, msg);
```

Comment:
To avoid some temporary string objects:
```
const std::string msg = StrCat(
    "planning has no trajectory point. planning_seq_num: ",
    local_view->trajectory.header().sequence_num());```